tools/nos_to_cc.c

```diff
@@ -0,0 +1,174 @@
+
+#if HAVE_CONFIG_H
+#include "../config.h"
+#endif
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "nereon.h"
+
+/*
+ * read file contents
+ */
+
+static size_t read_file_contents(const char *fpath, char **buf)
+{
+	FILE *fp;
+	struct stat st;
+	size_t read_len;
+
+	if (stat(fpath, &st) != 0 || !S_ISREG(st.st_mode) || st.st_size == 0) {
+		fprintf(stderr, "Could not get stat of file '%s'\n", fpath);
+		return 0;
+	}
+
+	fp = fopen(fpath, "r");
+	if (!fp) {
+		fprintf(stderr, "Could not open file '%s' for reading\n", fpath);
+		return 0;
+	}
+
+	*buf = (char *) malloc(st.st_size + 1);
+	if (*buf == NULL) {
+		fprintf(stderr, "Out of memory!\n");
+		fclose(fp);
+		return 0;
+	}
+
+	read_len = fread(*buf, 1, st.st_size, fp);
+	if (read_len > 0)
+		(*buf)[read_len] = '\0';
+
+	fclose(fp);
+
+	return read_len;
+}
+
+/*
+ * open file in safe writting mode
+ */
+
+FILE *fopen_wsafe(const char *fpath)
+{
+	int fd;
+	FILE *fp;
+
+	remove(fpath);
+
+	fd = open(fpath, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
+	if (fd < 0)
+		return NULL;
+
+	fp = fdopen(fd, "w");
+	if (!fp)
+		close(fd);
+
+	return fp;
+}
+
+/*
+ * write source and header file
+ */
+
+static int write_cc_files(const char *src_fpath, const char *hdr_fpath, const char *nos_str)
+{
+	FILE *src_fp, *hdr_fp;
+	const char *hdr_fname, *p;
+
+	/* get header file name */
+	p = strrchr(hdr_fpath, '/');
+	if (p)
+		hdr_fname = p + 1;
+	else
+		hdr_fname = hdr_fpath;
+
+	/* create source file */
+	src_fp = fopen_wsafe(src_fpath);
+	if (!src_fp)
+		return -1;
+
+	fprintf(src_fp, "#include \"%s\"\n\n", hdr_fname);
+	fprintf(src_fp, "static const char *g_nos_cfg = \"");
+
+	p = nos_str;
+	while (*p != '\0') {
+		if (*p == '"' || *p == '\n') {
+			fputc('\\', src_fp);
+		}
+		fputc(*p, src_fp);
+
```

Comment:
`\n` handling looks wrong to me - maybe something like:
```
if (*p == '\n')
    fprintf(src_fp, "\\n");
else {
    if (*p == '"')
        fputc("\\", src_fp);
    fputc(*p, strc_fp);
}
```

Reply:
Agree

Reply:
@ronaldtse @drystone Fixed!